### PR TITLE
FIX: extract and fix overriding of usernames by external auth

### PR DESCRIPTION
**Pull request by AndrewPrigorshnev**
This addresses comments of @gschlager to https://github.com/discourse/discourse/pull/14531:
- extracts the common logic of overriding usernames into separate method
- fixes usernames comparison when checking if a username should be overridden. We should unicode normalize a new username before comparison, not only down case it. Note also that this logic was extracted into a separate method:
  https://github.com/discourse/discourse/blob/78cfaa6f726c1fe08a6afd93c67caded2620548d/app/models/user.rb#L1441-L1443
- adds using of `UsernameChanger` instead of just setting a new username in order to update quotes and mentions when renaming (which addresses this complaint https://meta.discourse.org/t/updating-mentions-when-username-is-changed-via-sso/202992)

### CODE REVIEW

File: app/services/username_overrider.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class UsernameOverrider
+  def self.override(user, new_username)
```
[the remark this answers is not in the document]
Comment (AndrewPrigorshnev) [reply app/services/username_overrider.rb]: Yes, this makes sense, I was on the fence here. I've moved the method to `UsernameChanger`.

File: app/services/username_overrider.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    if user.username_equals_to?(new_username)
+      # override anyway since case could've been changed:
+      UsernameChanger.change(user, new_username, user)
```
[the remark this answers is not in the document]
Comment (AndrewPrigorshnev) [reply app/services/username_overrider.rb]: A good suggestion 👍, I've added the check to `UsernameChanger.change` and also added a test case.

File: spec/services/username_changer_spec.rb [hunk 3]
```diff
@@ -591,4 +603,70 @@ def username_and_something_else(username)
...
+
+    context "unicode_usernames is off" do
+      before do
+        SiteSetting.unicode_usernames = false
+      end
+
+      [
+        *common_test_cases,
+        [
+          "overrides the username if a new name after unicode normalization is different only in case",
```
Comment (gschlager) [inline spec/services/username_changer_spec.rb:630]: this test case description seems misleading given that `current` and `overrode` are both "john"

**Review by gschlager — CHANGES_REQUESTED** [review]

File: spec/services/username_changer_spec.rb (hunk 3 above)
Comment (AndrewPrigorshnev) [reply spec/services/username_changer_spec.rb]: @gschlager, I've renamed the test case, I probably copied it from another one and forgot to rename it.

Comment (AndrewPrigorshnev) [reply spec/services/username_changer_spec.rb]: Oops, sorry, I haven't seen it's already approved. Thank you!

**Review by ZogStriP — APPROVED** [review]

**Pull request merged**